File: repl/adrianus/cpp/thread_utils.cpp
#include "thread_utils.h"
#include "thread_workers.h"
#include "objects.h"
#include "gc.h"
#include <chrono>
#include <thread>
#include <iostream>
#include "gc.h"
#include "environment.h"

void test() {
    std::cout << "start test thread\n";
    for (int i = 0; i < 100; i++) {
        std::cout << i << "\n";
    }
    std::cout << "finish test thread\n";
    TOTAL_THREADS_RUNNING--;
}

void thread_callback(Ad_Object* rawObject, std::vector<Ad_Object*> arg_objs) {
    Ad_Thread_Object *threadObject = (Ad_Thread_Object*) rawObject;

    GarbageCollector *gc2 = new GarbageCollector();
    threadObject->internal_gc = gc2;

    threadObject->callback = arg_objs.at(0);
    std::vector<Ad_Object*> params;
    for (int i = 1; i < arg_objs.size(); i++) {
        params.push_back(arg_objs.at(i)->copy(threadObject->internal_gc));
    }
    threadObject->params = params;
}

void thread_async_run(Ad_Object* rawObject, GarbageCollector *gc, Environment &env) {
    Ad_Thread_Object* threadObject = (Ad_Thread_Object*) rawObject;
    TOTAL_THREADS_RUNNING++;
    std::thread *th1 = new std::thread(ad_worker_async, threadObject->callback, threadObject->params, threadObject->internal_gc, &env);
    threadObject->internal_thread = th1;
    threadPool.push_back(threadObject);
}

void thread_blocking_run(Ad_Object* rawObject, GarbageCollector *gc, Environment &env) {
    Ad_Thread_Object* threadObject = (Ad_Thread_Object*) rawObject;
    // TODO: better structure for this
    TOTAL_THREADS_RUNNING++;
    std::thread *th1 = new std::thread(ad_worker_blocking, threadObject->callback, threadObject->params, threadObject->internal_gc, &env);
    threadObject->internal_thread = th1;
    threadPool.push_back(threadObject);
    th1->join();
}

void sleep_builtin_executor(int value) {
    std::this_thread::sleep_for (std::chrono::milliseconds(value));
}

File: repl/adrianus/python3/thread_utils.py
import time
import threading
from thread_worker import ad_worker_async, ad_worker_blocking

def thread_callback(thread_obj, arg_objs):
    thread_obj.callback = arg_objs[0]
    thread_obj.params = arg_objs[1:]

def thread_async_run(thread_obj, env):
    th1 = threading.Thread(target=ad_worker_async, args=(thread_obj.callback, thread_obj.params, thread_obj, env))
    thread_obj.thread = th1
    th1.start()

def thread_blocking_run(thread_obj, env):
    th1 = threading.Thread(target=ad_worker_blocking, args=(thread_obj.callback, thread_obj.params, thread_obj, env))
    thread_obj.thread = th1
    th1.start()
    th1.join()

def thread_await(thread_obj, env):
    thread_obj.thread.join()
    return thread_obj.result

def sleep_builtin_executor(value):
    time.sleep(value / 1000)
